daft/runners/ray_runner.py

```diff
@@ -695,121 +846,80 @@ def place_in_queue(item):
                 except Full:
                     pass
 
-        with profiler(profile_filename):
-            try:
-                next_step = next(tasks)
+        with profiler(profile_filename), ray_tracing.ray_tracer(result_uuid) as runner_tracer:
+            raw_tasks = plan_scheduler.to_partition_tasks(
+                psets,
+                self,
+                # Attempt to subtract 1 from results_buffer_size because the return Queue size is already 1
+                # If results_buffer_size=1 though, we can't do much and the total buffer size actually has to be >= 2
+                # because we have two buffers (the Queue and the buffer inside the `materialize` generator)
+                None if results_buffer_size is None else max(results_buffer_size - 1, 1),
+            )
+            tasks = ray_tracing.MaterializedPhysicalPlanWrapper(raw_tasks, runner_tracer)
 
+            try:
+                wave_count = 0
                 while is_active():  # Loop: Dispatch -> await.
-                    while is_active():  # Loop: Dispatch (get tasks -> batch dispatch).
-                        tasks_to_dispatch: list[PartitionTask] = []
-
-                        # TODO: improve control loop code to be more understandable and dynamically adjust backlog
-                        cores: int = max(
-                            next(num_cpus_provider) - self.reserved_cores, 1
-                        )  # assume at least 1 CPU core for bootstrapping clusters that scale from zero
-                        max_inflight_tasks = cores + self.max_task_backlog
-                        dispatches_allowed = max_inflight_tasks - len(inflight_tasks)
-                        dispatches_allowed = min(cores, dispatches_allowed)
-
-                        # Loop: Get a batch of tasks.
-                        while len(tasks_to_dispatch) < dispatches_allowed and is_active():
-                            if next_step is None:
-                                # Blocked on already dispatched tasks; await some tasks.
+                    wave_count += 1
+                    with runner_tracer.dispatch_wave(wave_count):
```

Comment:
I refactored our code for the dispatch loop into 3 main helpers:

1. `self._construct_dispatch_batch`
2. `self._dispatch_tasks`
3. `self._await_tasks`

This was done because adding tracing on top of the current heavily nested while loop made the code extremely bug-prone (adding even more nesting with the tracing `with` statements)

The existing behavior of our dispatch loop should be well-maintained, with the exception of a change I made to where iteration of the PhysicalPlan is performed: https://github.com/Eventual-Inc/Daft/pull/3113/files#r1818218689